udf/num_weeks_gte_2_from_bits.sql

```diff
@@ -0,0 +1,12 @@
+CREATE OR REPLACE FUNCTION
+  udf.num_weeks_gte_n_from_bits(b BYTES, i INT64) AS (
+    CAST(
+        BIT_COUNT(b & udf.bitmask_range(1, 7)) >= i AS INT64
+    ) + CAST(
+        BIT_COUNT(b & udf.bitmask_range(8, 7)) >= i AS INT64
+    ) + CAST(
+        BIT_COUNT(b & udf.bitmask_range(15, 7)) >= i AS INT64
+    ) + CAST(
+        BIT_COUNT(b & udf.bitmask_range(22, 7)) >= i AS INT64
```

Comment:
The 28 bits we have are generally thought to represent the current day (day 0) backward in time to day 27. Week 0 is days 0 through 6, etc. so that Week 3 is days 20 through 27.

So I would expect the offsets here to be 0, 7, 14, and 21 rather than 1, 8, 15, and 22.

Reply:
That is how we've been thinking about retention for GUD.

Reply:
When we partition clients into segments to study their data for some day, that day's data _shouldn't_ be used to determine which segment the client falls into. So if we want to look at today's DAU among "regular users v1", and today is day 0, then looking backwards we need to look at days 1 through 7, etc. Otherwise our segments can become trivial - "users who are highly engaged tend to be highly engaged" vs "users who have been highly engaged tend to stay highly engaged".

So, either in the analysis or the ETL we need to look one day backward from the day of interest.

I _think_ it's best to do this here in the ETL, because people using `clients_last_seen` to compute DAU would get the right behaviour from the naive query. People wanting to compute WAU or MAU, and people wanting to use other tables, would need to `LEFT JOIN clients_last_seen` and query the row from the correct `submission_date`, the date that the client first appears in their analysis.

Happy to discuss this point further via other channels if it's not clear here; it's an important subtlety and I would love to be surprised by a solution cleaner than what I've proposed.

Reply:
Wait, I'm wrong - `udf.bitmask_range(22, 7)` goes off the back of the array; we do the "look one day backward" by using `_previous.days_visited_5_uri_bits` rather than `_current.days_visited_5_uri_bits`. I'll update it.

Reply:
...ah, I was originally right. `udf.bitmask_range` internally subtracts 1 from its input, raises an error when I try `udf.bitmask_range(0, 7)` and has a test case `assert_equals(1, udf.bitmask_range(1, 1))`

Reply:
Sorry for the confusion there. Indeed, this function uses ordinal indices to match the SUBSTR function and I had forgotten that detail.